### PR TITLE
Introduce new parameter Expected Failures List

### PR DESCRIPTION
**Pull request by kstrenkova**
The Expected Failures List parameter will help with specifying set of tests whose failures we want to ignore. By adding test to this list, its failure will not have impact on the jobs pass/fail. It can help with not having to re-run long running jobs and save time and resources in specific cases, for example when dealing with flaky tests.

### CODE REVIEW

File: config/crd/bases/test.openstack.org_tempests.yaml [hunk 1]
```diff
@@ -168,6 +168,10 @@ spec:
                     description: A content of exclude.txt file that is passed to tempest
                       via --exclude-list
                     type: string
+                  expectedFailuresList:
+                    description: A content of expected_failures.txt file that is passed
+                      to tempest via --expected-failures-list
```
Comment (lpiwowar) [inline config/crd/bases/test.openstack.org_tempests.yaml:173]: @kopecmartin do you think `--expected-failures-list` is something that would be "mergeable" in tempest.

Comment (kopecmartin) [reply config/crd/bases/test.openstack.org_tempests.yaml:173]: this doesn't belong to upstream tempest, tempest tests are meant to be explicit - https://opendev.org/openstack/tempest/src/branch/master/REVIEWING.rst#being-explicit
Parameter like this may mask failures (especially if a user forgets that a test result is being ignored) because they'll be ignored .. if tests fail due to a known bug, the user should not run them - if it's a bug in tempest, it can be skipped in tempest, if it's a bug in a component, well, the test did its job - if a user doesn't want a failing CI in the meantime, they can exclude such test.

that's an upstream perspective ^^

Downstream perspective:
I'm still not convinced this is really needed as any failing or potentially failing test can be simply excluded.

Comment (arxcruz) [reply config/crd/bases/test.openstack.org_tempests.yaml:173]: This have nothing to do with tempest in upstream. I expect to have a list of tests that I know it's unstable (timeout, concurrency issues, etc) and I want testoperator to check if the match list of expected failing tests matches the failing tests in a particular job, and if does, don't mark the job as failure. The tempest test will continue to fail, and that is okay, but in CI, the job will be marked as success and we have the failures reported to report portal/polarion. This is important to add unijobs into the promotion criteria, since unijobs run 5k tests, sometimes a test fails because of concurrency, or whitebox tests for example that are very unstable. But we want to collect the data of how many times it fails/pass so we can decide later what to do with these tests.

Comment (kopecmartin) [reply config/crd/bases/test.openstack.org_tempests.yaml:173]: all clear now, this in test-operator is gonna be an interface how to get the tests (that might fail) to tcib script where the "don't fail on these tests" logic will be written:
https://github.com/openstack-k8s-operators/tcib/blob/main/container-images/tcib/base/os/tempest/run_tempest.sh

this won't reach the tempest upstream.

Comment (arxcruz) [reply config/crd/bases/test.openstack.org_tempests.yaml:173]: Exactly, this is a unijobs downstream feature only. Nothing related to upstream tempest.

Comment (lpiwowar) [reply config/crd/bases/test.openstack.org_tempests.yaml:173]: Ok, it seems we have an agreement and common understanding of how it should work. In this case just Katka, please, when you have time and this PR is ready, update the description of the `expectedFailuresList`. Right now it suggests that the `--expected-failures-list` parameter is passed to tempest.

Comment (kstrenkova) [reply config/crd/bases/test.openstack.org_tempests.yaml]: Thank you @arxcruz for making the intention for this parameter clear. It is my mistake for writing the description in such a way because I followed the code in include and exclude list. I fixed the description in the last push, so I hope from now on it will be smooth sailing. Sorry for the confusion :sweat_smile:

Comment (lpiwowar) [reply config/crd/bases/test.openstack.org_tempests.yaml]: That's ok, @kstrenkova. At least we sorted things out :)

**Review by lpiwowar — CHANGES_REQUESTED** [review]
Just two small request otherwise looks good to me!:) :+1:

File: api/v1beta1/tempest_types.go [hunk 2]
```diff
@@ -151,6 +151,12 @@ type TempestRunSpec struct {
 	// A content of exclude.txt file that is passed to tempest via --exclude-list
 	ExcludeList string `json:"excludeList"`
 
+	// +kubebuilder:validation:Optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec
+	// The expected_failures.txt file contains tests that should not count
+	// as failures for the job, as they are expected to fail.
```
[the remark this answers is not in the document]
Comment (kstrenkova) [reply api/v1beta1/tempest_types.go]: I agree with you :smile: I just did a minor fixes to the sentence.

File: config/samples/test_v1beta1_tempest.yaml [hunk 3]
```diff
@@ -46,6 +46,8 @@ spec:
     concurrency: 8
     # excludeList: | # <-- Use | to preserve \n
     #   tempest.api.identity.v3.*
+    # expectedFailuresList: | # <-- Use | to preserve \n
+    #   tempest.api.identity.v3.*
```
[the remark this answers is not in the document]
Comment (kstrenkova) [reply config/samples/test_v1beta1_tempest.yaml]: Okay, that makes sense. Thanks for pointing that out :+1:

**Comment by kstrenkova** [comment]
I have make the requested changes and changed the definition of the `podified-multinode-edpm-deployment-crc-test-operator` job, so it uses newly introduced parameter.

File: .zuul.yaml [hunk 4]
```diff
@@ -57,6 +57,7 @@
         tempest.api.identity.admin.v3.test_credentials.CredentialsTestJSON
         tempest.api.identity.admin.v3.test_tokens.TokensV3TestJSON.test_rescope_token
         tempest.api.identity.admin.v3.test_users.UsersV3TestJSON.test_update_user_password
+      cifmw_test_operator_tempest_expected_failures_list: |
```
Comment (lpiwowar) [inline .zuul.yaml:60]:
```suggestion
      cifmw_test_operator_tempest_expected_failures_list: |
        foobar
```

**Review by lpiwowar — CHANGES_REQUESTED** [review]
You can also add Depends-On: https://github.com/openstack-k8s-operators/ci-framework/pull/2513

We can test the ci-framework change together with this PR as well:).

**Comment by lpiwowar** [comment]
> I have make the requested changes and changed the definition of the `podified-multinode-edpm-deployment-crc-test-operator` job, so it uses newly introduced parameter.

thanks! :tada:

File: .zuul.yaml (hunk 4 above)
Comment (kstrenkova) [reply .zuul.yaml:60]: Yeah, I was curious if the option can be set empty like this. Do I need to add foobar for it to work? :thinking:

Comment (lpiwowar) [reply .zuul.yaml:60]: You could leave it empty but then it would be better to define the var like this (no need for `|` [1]).

```
cifmw_test_operator_tempest_expected_failures_list: ""
``` 
But I would prefer if you added the `foobar` value. Think about it in terms of code coverage. How much lines of the newly added code to test-operator and tcib would be executed if you lave the value stored in the variable empty in contrast with the `foobar` value? :thinking: [2]

It's definitely good to testy for edge cases like empty value but in this job it would be good if we tried to cover as much code as possible.

[1] [here](https://docs.ansible.com/ansible/latest/reference_appendices/YAMLSyntax.html#yaml-basics)
[2] [here](https://github.com/openstack-k8s-operators/test-operator/pull/233/files#diff-19fb6c8a471eb8f571e5e959d274704a582df8fa55a60f500593fcba5efb8823R448)

Comment (kstrenkova) [reply .zuul.yaml:60]: Alright, thank you for your answer. That was an interesting view on this, I did not think about it in that way :smile:

**Review by lpiwowar — COMMENTED** [review]
Looks good to me!:) :+1:  Let's wait for the tcib change to get merged first.

P.S.: If you want to improve a little bit the description of the parameter then I think it would be great to mention that the parameter does not accept regexes but only exact full test names.

**Comment by lpiwowar** [comment]
recheck

**Comment by lpiwowar** [comment]
recheck
